### PR TITLE
Add metadata function

### LINKED ISSUE #94: Add file size and type to TrashItem
When visualizing the items in the trash, it is helpful to know if the item is a file or a folder, and what size it is. I looked into both Freedesktop and Windows implementations for this, and it seems to be possible in both cases. For Linux, just getting the metadata of the file or folder in the trash is enough. For Windows, the IShellItem has a GetAttributes method that returns enough information to distinguish files from folders: https://learn.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ishellitem-getattributes. I'm sure a similar method exists for getting the size of files in the Recycle Bin.

If desired I can work on an implementation.

### PR DESCRIPTION
**Pull request by jackpot51**
Resolves #94, adds a metadata function to get the type of the item (file or folder) and the size (in bytes for files or entries for folders). This is implemented for Freedesktop platforms and Windows

### CODE REVIEW

**Comment by jackpot51** [comment]
I've also added a stub to allow unix platforms that aren't yet supported by get_mount_points to at least work with the home trash.

File: src/lib.rs [hunk 1]
```diff
@@ -304,6 +304,15 @@ impl Hash for TrashItem {
     }
 }
 
+/// Metadata about a [`TrashItem`]
+#[derive(Debug, Clone)]
+pub struct TrashItemMetadata {
+    /// True if the [`TrashItem`] is a directory, false if it is a file
+    pub is_dir: bool,
+    /// Number of entries for directories, number of bytes for files
+    pub size: u64,
```
Comment (Byron) [inline src/lib.rs:313]: I think this type makes it too easy to mis-use `size` and treat it as a size of bytes.
Instead, an `enum` could make it absolutely clear what is what, with field names that reflect what's counted.

That enum could then have convenience methods to get the size in bytes for example as `Option<u64>`, and maybe the same for the entries in a directory. Speaking of, I would make clear that this count is non-recursive.

File: src/windows.rs [hunk 2]
```diff
@@ -124,6 +127,43 @@ pub fn list() -> Result<Vec<TrashItem>, Error> {
     }
 }
 
+pub fn metadata(item: &TrashItem) -> Result<TrashItemMetadata, Error> {
+    ensure_com_initialized();
+    unsafe {
+        let id_as_wide: Vec<u16> = item.id.encode_wide().chain(std::iter::once(0)).collect();
```
Comment (Byron) [inline src/windows.rs:133]: I don't want to burden you, but I'd love it if there was a `to_wide_path(&OsStr) -> Vec<u16>` or something that replaces this code here and in 5 other places in this file.

File: src/windows.rs [hunk 3]
```diff
@@ -124,6 +127,43 @@ pub fn list() -> Result<Vec<TrashItem>, Error> {
     }
 }
 
+pub fn metadata(item: &TrashItem) -> Result<TrashItemMetadata, Error> {
+    ensure_com_initialized();
+    unsafe {
```
Comment (Byron) [inline src/windows.rs:132]: Could each unsafe function call get its own unsafe block? I understand that this is how most of the functions are done, but I don't want this practice to proliferate.
It's just a small step towards better `unsafe`, as I still wouldn't be able to reason about safety here 😅 (*nor do I have the expectation that this needs to happen here*).

For context, I [recently have been shown how to deal with `unsafe`](https://github.com/Byron/gitoxide/pull/1237), and I think it affected me 😁.

**Review by Byron — CHANGES_REQUESTED** [review]
Thanks a lot for all the work you put in to make this happen! 

Particularly Windows looks like it wasn't easy to get there (and it feels like one has to read a book to do anything on this platform).

The biggest change request is probably around how metadata is represented, and I hope it makes sense to you as well.

**Comment by jackpot51** [comment]
Thanks for the review! I will work on addressing these tomorrow

File: src/freedesktop.rs [hunk 4]
```diff
@@ -808,6 +808,18 @@ fn get_mount_points() -> Result<Vec<MountPoint>, Error> {
...
 }
 
+#[cfg(not(any(
+    target_os = "linux",
+    target_os = "dragonfly",
+    target_os = "freebsd",
+    target_os = "openbsd",
+    target_os = "netbsd"
+)))]
+fn get_mount_points() -> Result<Vec<MountPoint>, Error> {
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/freedesktop.rs:818]: I will return an error instead.

File: src/freedesktop.rs [hunk 5]
```diff
@@ -218,6 +218,23 @@ pub fn list() -> Result<Vec<TrashItem>, Error> {
...
+pub fn metadata(item: &TrashItem) -> Result<TrashItemMetadata, Error> {
+    // When purging an item the "in-trash" filename must be parsed from the trashinfo filename
+    // which is the filename in the `id` field.
+    let info_file = &item.id;
+
+    // A bunch of unwraps here. This is fine because if any of these fail that means
+    // that either there's a bug in this code or the target system didn't follow
+    // the specification.
+    let file = restorable_file_in_trash_from_info_file(info_file);
+    assert!(virtually_exists(&file).map_err(|e| fs_error(&file, e))?);
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/freedesktop.rs:230]: I will address this in the next commit

File: src/freedesktop.rs [hunk 6]
```diff
@@ -218,6 +218,23 @@ pub fn list() -> Result<Vec<TrashItem>, Error> {
     Ok(result)
 }
 
+pub fn metadata(item: &TrashItem) -> Result<TrashItemMetadata, Error> {
+    // When purging an item the "in-trash" filename must be parsed from the trashinfo filename
+    // which is the filename in the `id` field.
+    let info_file = &item.id;
+
+    // A bunch of unwraps here. This is fine because if any of these fail that means
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/freedesktop.rs:226]: I will address this in the next commit

File: src/lib.rs (hunk 1 above)
Comment (jackpot51) [reply src/lib.rs:313]: I will address this in the next commit

File: src/windows.rs (hunk 3 above)
Comment (jackpot51) [reply src/windows.rs:132]: I will address this in the next commit

**Comment by jackpot51** [comment]
@Byron I addressed most of your comments, please take a look.

**Comment by Byron** [comment]
Thanks a lot for addressing most of the changes and for contributing, it's appreciated.

I did the rest of the changes and will merge in a moment.

**Pull request merged**